doc/man/align_detector.1.md
% align_detector(1)

NAME
====

align_detector - refine detector geometry


SYNOPSIS
========

align_detector -g _input.geom_ -o _output.geom_ -l _level_ [--out-of-plane] _millepede-files_


DESCRIPTION
===========

**align_detector** refines the detector geometry based on the calibration data
written by **indexamajig**.  The refinement takes into account all the
inter-dependencies between crystal orientations, cell parameters and the panel
positions, as if a single large minimisation had been performed with all frames
at once.  The algorithm is nevertheless very fast - this is achieved using the
Millepede-II algorithm.  For more information, see
https://www.desy.de/~kleinwrt/MP2/doc/html/index.html

To refine the detector geometry, first make sure that the geometry file
includes hierarchy information.  For this, see **man crystfel_geometry**,
section **Detector hierarchy**.

Next, run **indexamajig** as usual, but with option **--mille**.  This will
produce an additional output file with default name **mille-data.bin** (use
**--mille-file** to set a different location).  This file contains calibration
data that can be read by **align_detector** and **show_residuals**.

Finally, run **align_detector**, giving it the input geometry file, the "Mille"
files, a refinement level and a filename for the updated geometry file.  The
input geometry file must match the file used for the indexamajig run.

Refinement level **0** allows only the overall detector position to vary.
Higher levels allow groups of panels to move according to the hierarchy.  For
example, for a CSPAD detector with hierarchy defined as in the example files,
level **1** would refine the quadrant positions, **2** would refine the 2-by-1
panel "dominoes", and level **3** would refine the individual (roughly square)
ASICs.  Higher refinement levels will generally require more and
higher-resolution diffraction data.

You can use **indexamajig** option **--max-mille-level** to reduce the size of
the calibration data files, at the cost of limiting the maximum refinement
level.

The default behaviour is to refine only the position components in the x-y
plane, perpendicular to the beam.  In favourable circumstances, you can add
option **--out-of-plane** to refine the panel **translations** out of this
plane. However, be aware that this introduces additional cross-dependencies
and is less stable.  Add option **--out-of-plane-tilts** to additionally
refine the panel **tilts** out of the x-y plane, which is even less stable.

The overall detector z-position (camera length) is usually quite strongly
correlated with the crystal cell parameters, making it difficult to refine.
With high-resolution data, it can be possible, however.  To refine the overall
camera length, add **--camera-length**.  Without this option, the overall
camera length will not be altered, even with **--out-of-plane**.  Instead, only
the z-positions of the panels relative to one another will be refined.

**align_detector** relies on the program **pede** from the Millepede-II
package.  Usually, this will be installed as part of the CrystFEL installation
procedure.  If not, it can easily be installed from the Millepede-II repository
at https://gitlab.desy.de/claus.kleinwort/millepede-ii


OPTIONS
=======

**-g** _input.geom_, **--geometry**=_input.geom_
: Specify the input geometry filename.

**-o** _output.geom_, **--output**=_output.geom_
: Specify the output geometry filename.
: Note that the geometry file will be re-written, meaning that any formatting
: and comments will be lost.

**-l** _level_, **--level**=_level_
: Specify the refinement level.  **-l 0** refines the overall detector position
: only.  The maximum refinement level is determined by the hierarchy of the
: detector.

**--out-of-plane**
: Additionally refine out-of-plane panel translations, relative to one another
: (see **--camera-length**).

**--out-of-plane-tilts**
: Additionally refine out-of-plane panel tilts.

**--camera-length**
: Additionally refine the overall camera length.

**--panel-totals**
: Display the total translation and rotation changes for each panel. This does
: not affect the refinement or the output geometry file in any way, only the
: console output.

**--min-measurements**=_n_
: Require at least _n_ measurements contributing to a detector geometry
: parameter before updating that parameter.  Higher values give more stable
: refinement.  Note that there are two measurements per peak-reflection pair,
: because the position offsets are two dimensional.  The default value is
: 100, i.e. at least 50 pairs are needed for each parameter.

**--big-errors**
: Tell **pede** to be very tolerant of large residuals in the input data.
: Add this option if an initial refinement fails due to too many "huge"
: rejects.


AUTHOR
======

This page was written by Thomas White.


REPORTING BUGS
==============

Report bugs to <taw@physics.org>, or visit <http://www.desy.de/~twhite/crystfel>.


COPYRIGHT AND DISCLAIMER
========================

Copyright © 2023-2025 Deutsches Elektronen-Synchrotron DESY, a research centre
of the Helmholtz Association.

align_detector, and this manual, are part of CrystFEL.

CrystFEL is free software: you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

CrystFEL is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
CrystFEL.  If not, see <http://www.gnu.org/licenses/>.


SEE ALSO
========

**crystfel**(7), **indexamajig**(1), **adjust_detector**(1), **show_residuals**(1)
**crystfel_geometry**(5)


NEWS
There is no NEWS.  Check the version control repository...


src/align_detector.c
/*
 * align_detector.c
 *
 * Align detector using Millepede
 *
 * Copyright © 2023-2026 Deutsches Elektronen-Synchrotron DESY,
 *                       a research centre of the Helmholtz Association.
 *
 * Authors:
 *   2023-2026 Thomas White <taw@physics.org>
 *
 * This file is part of CrystFEL.
 *
 * CrystFEL is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * CrystFEL is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with CrystFEL.  If not, see <http://www.gnu.org/licenses/>.
 *
 */


#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <unistd.h>
#include <dirent.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <getopt.h>
#include <libgen.h>

#include <datatemplate.h>
#include <utils.h>
#include <predict-refine.h>
#include <crystfel-mille.h>

#include "version.h"


static void show_syntax(const char *s)
{
	printf("Syntax: %s [options] -g <input.geom> -o <output.geom> <mille-0.dat> [...]\n", s);
}


static void show_help(const char *s)
{
	show_syntax(s);
	printf("\nRefine detector geometry using Millepede.\n"
	       "\n"
	       "  -g, --geometry=file        Input geometry file\n"
	       "  -o, --output=file          Output geometry file\n"
	       "  -l, --level=n              Alignment hierarchy level\n"
	       "      --out-of-plane         Refine (relative) z-direction shifts\n"
	       "      --out-of-plane-tilts   Refine panel rotations around x and y\n"
	       "      --camera-length        Refine overall camera length\n"
	       "      --panel-totals         Display total panel movements\n"
	       "      --min-measurements     Minimum measurements per parameter\n"
	       "      --big-errors           Increase tolerance for large residuals\n"
	       "\n"
	       "  -h, --help                 Display this help message\n"
	       "      --version              Print version number and exit\n");
}


static const char *str_param(enum gparam param)
{
	switch ( param ) {
		case GPARAM_DET_TX : return "x-translation";
		case GPARAM_DET_TY : return "y-translation";
		case GPARAM_DET_TZ : return "z-translation";
		case GPARAM_DET_RX : return "x-rotation";
		case GPARAM_DET_RY : return "y-rotation";
		case GPARAM_DET_RZ : return "z-rotation";
		case GPARAM_SCANV_BEAM_A : return "x/scanX";
		case GPARAM_SCANV_BEAM_B : return "x/scanY";
		case GPARAM_SCANV_BEAM_C : return "y/scanX";
		case GPARAM_SCANV_BEAM_D : return "y/scanY";
		default : return "(unknown)";
	}
}


static const char *group_serial_to_name(int serial,
                                        struct dg_group_info *groups,
                                        int n_groups)
{
	int i;

	for ( i=0; i<n_groups; i++ ) {
		if ( groups[i].serial == serial ) return groups[i].name;
	}

	return NULL;
}


static struct dg_group_info *find_group(struct dg_group_info *groups,
                                        int n_groups, const char *name)
{
	int i;

	for ( i=0; i<n_groups; i++ ) {
		if ( strcmp(name, groups[i].name) == 0 ) return &groups[i];
	}

	return NULL;
}

static int ipow(int base, int ex)
{
	int i;
	int v = 1;
	for ( i=0; i<ex; i++ ) {
		v *= base;
	}
	return v;
}


static int is_child(struct dg_group_info *parent, struct dg_group_info *child)
{
	int parent_serial;

	if ( 1+parent->hierarchy_level != child->hierarchy_level ) return 0;

	parent_serial = child->serial % ipow(100, child->hierarchy_level);
	if ( parent->serial != parent_serial ) return 0;

	return 1;
}


static void write_zero_sum(FILE *fh, struct dg_group_info *g,
                           struct dg_group_info *groups, int n_groups,
                           enum gparam p)
{
	int i;
	int found = 0;

	for ( i=0; i<n_groups; i++ ) {
		if ( is_child(g, &groups[i]) ) {
			if ( !found ) {
				fprintf(fh, "Constraint 0\n");
				found = 1;
			}
			fprintf(fh, "%i 1\n", mille_label(groups[i].serial, p));
		}
	}

	if ( found ) {
		fprintf(fh, "\n");
	}
}


static int make_zero_sum(FILE *fh, struct dg_group_info *groups, int n_groups,
                         const char *group_name, int level,
                         int out_of_plane_shift, int out_of_plane_tilts)
{
	int i;
	struct dg_group_info *g = find_group(groups, n_groups, group_name);

	if ( g == NULL ) {
		ERROR("Couldn't find group '%s'\n", group_name);
		return 1;
	}

	/* Millepede doesn't like excessive constraints */
	if ( g->hierarchy_level >= level ) return 0;

	fprintf(fh, "! Hierarchy constraints for group %s\n", group_name);
	write_zero_sum(fh, g, groups, n_groups, GPARAM_DET_TX);
	write_zero_sum(fh, g, groups, n_groups, GPARAM_DET_TY);
	if ( out_of_plane_shift ) {
		write_zero_sum(fh, g, groups, n_groups, GPARAM_DET_TZ);
	}
	if ( out_of_plane_tilts ) {
		write_zero_sum(fh, g, groups, n_groups, GPARAM_DET_RX);
		write_zero_sum(fh, g, groups, n_groups, GPARAM_DET_RY);
	}
	write_zero_sum(fh, g, groups, n_groups, GPARAM_DET_RZ);
	fprintf(fh, "\n");

	for ( i=0; i<n_groups; i++ ) {
		if ( is_child(g, &groups[i]) ) {
			if ( make_zero_sum(fh, groups, n_groups, groups[i].name,
			                   level, out_of_plane_shift,
			                   out_of_plane_tilts) ) return 1;
		}
	}

	return 0;
}


static void print_time_warning()
{
	ERROR("\n\n");
	ERROR("WARNING: The modification times of the Mille data files differ "
	      "by more than one minute.\n");
	ERROR("This suggests that they might have come from different runs of "
	      "indexamajig.\n");
	ERROR("If this is not the case, then there is no problem and you can "
	      "ignore this warning.\n");
	ERROR("However, if Mille files have been mixed up between indexamajig "
	      "runs, then the detector alignment results will be wrong.\n");
	ERROR("\n\n");
}


static int different(time_t a, time_t b)
{
	if ( a-b > 60 ) return 1;
	if ( b-a > 60 ) return 1;
	return 0;
}


static int run_pede()
{
	char buf[1024];
	char cmdline[1280];
	int r;
	FILE *fh;

	if ( readlink("/proc/self/exe", buf, 1024) != -1 ) {
		char *dir = dirname(buf);
		strcpy(cmdline, dir);
		strcat(cmdline, "/pede");
		if ( !file_exists(cmdline) ) {
			ERROR("Couldn't find %s - falling back on shell path\n",
			      cmdline);
			strcpy(cmdline, "pede millepede.txt");
		} else {
			strcat(cmdline, " millepede.txt");
		}
	} else {
		ERROR("readlink() failed: %s\n", strerror(errno));
		strcpy(cmdline, "pede millepede.txt");
	}

	STATUS("Running '%s'\n", cmdline);
	r = system(cmdline);
	if ( r == -1 ) {
		ERROR("Failed to run Millepde: %s\n", strerror(errno));
		return 1;
	}
	if ( !WIFEXITED(r) ) {
		ERROR("Millepede exited abnormally.\n");
		return 1;
	}
	if ( WEXITSTATUS(r) != 0 ) {
		ERROR("Millepede returned an error status (%i)\n", WEXITSTATUS(r));
		return 1;
	}

	fh = fopen("millepede.end", "r");
	if ( fh == NULL ) {
		ERROR("Failed to open millepede.end\n");
		return 1;
	}

	if ( fgets(buf, 256, fh) != buf ) {
		ERROR("Failed to read first line of millepede.end\n");
		return 1;
	}

	fclose(fh);

	chomp(buf);
	if ( sscanf(buf, "%i", &r) != 1 ) {
		ERROR("Failed to parse millepede.end: %s\n", buf);
		return 1;
	}

	STATUS("Millepede status: %s\n", buf);

	/* "Ended with severe warnings", or worse? */
	if ( r > 1 ) {
		ERROR("Status is %i, Millepede failed.\n", r);
		return 1;
	}

	return 0;
}


static void write_children(FILE *fh, const char *dir)
{
	DIR *d = opendir(dir);
	struct dirent *dent = readdir(d);

	while ( dent != NULL ) {
		if ( dent->d_name[0] == 'm' ) {
			fprintf(fh, "%s/%s\n", dir, dent->d_name);
		}
		dent = readdir(d);
	}

	closedir(d);
}


int main(int argc, char *argv[])
{
	int c;
	char *in_geom = NULL;
	char *out_geom = NULL;
	int level = 0;
	char *rval;
	int i;
	FILE *fh;
	DataTemplate *dtempl;
	struct dg_group_info *groups;
	int n_groups;
	int r;
	char line[256];
	time_t first_mtime = 0;
	int warn_times = 0;
	int out_of_plane_shift = 0;
	int out_of_plane_tilts = 0;
	int refine_clen = 0;
	int panel_totals = 0;
	int min_measurements = 100;
	int big_errors = 0;

	/* Long options */
	const struct option longopts[] = {

		{"help",               0, NULL,               'h'},
		{"verbose",            0, NULL,               'v'},

		{"version",            0, NULL,               'V'},
		{"input",              1, NULL,               'g'},
		{"output",             1, NULL,               'o'},
		{"level",              1, NULL,               'l'},
		{"out-of-plane",       0, &out_of_plane_shift, 1},
		{"out-of-plane-tilts", 0, &out_of_plane_tilts, 1},
		{"camera-length",      0, &refine_clen,        1},
		{"panel-totals",       0, &panel_totals,       1},
		{"big-errors",         0, &big_errors,         1},
		{"min-measurements",   1, NULL,                2},

		{0, 0, NULL, 0}
	};

	/* Short options */
	while ((c = getopt_long(argc, argv, "hVo:g:i:l:",
	                        longopts, NULL)) != -1)
	{

		switch (c) {

			case 'h' :
			show_help(argv[0]);
			return 0;

			case 'V' :
			printf("CrystFEL: %s\n", crystfel_version_string());
			printf("%s\n", crystfel_licence_string());
			return 0;

			case 'g' :
			case 'i' :
			in_geom = strdup(optarg);
			break;

			case 'o' :
			out_geom = strdup(optarg);
			break;

			case 'l' :
			errno = 0;
			level = strtol(optarg, &rval, 10);
			if ( *rval != '\0' ) {
				ERROR("Invalid value for --level.\n");
				return 1;
			}
			break;

			case 2 :
			errno = 0;
			min_measurements = strtol(optarg, &rval, 10);
			if ( *rval != '\0' ) {
				ERROR("Invalid value for --min-measurements.\n");
				return 1;
			}
			break;

			case 0 :
			break;

			case '?' :
			break;

			default :
			ERROR("Unhandled option '%c'\n", c);
			break;

		}

	}

	if ( (in_geom == NULL) || (out_geom == NULL) || (argc == optind) ) {
		show_syntax(argv[0]);
		return 1;
	}

	fh = fopen("millepede.txt", "w");
	if ( fh == NULL ) {
		ERROR("Couldn't open Millepede steering file\n");
		return 1;
	}

	for ( i=optind; i<argc; i++ ) {

		struct stat statbuf;

		r = stat(argv[i], &statbuf);
		if ( r != 0 ) {
			ERROR("File/directory '%s' not found\n", argv[i]);
			return 1;
		}

		if ( i == optind ) {
			first_mtime = statbuf.st_mtime;
		} else {
			if ( different(statbuf.st_mtime, first_mtime) ) {
				warn_times = 1;
			}
		}

		if ( is_dir(argv[i]) ) {
			write_children(fh, argv[i]);
		} else {
			fprintf(fh, "%s\n", argv[i]);
		}
	}

	if ( warn_times ) print_time_warning();

	dtempl = data_template_new_from_file(in_geom);
	groups = data_template_group_info(dtempl, &n_groups);

	fprintf(fh, "\nParameter\n");

	/* Top level */
	fprintf(fh, "%i 0 0\n", mille_label(0, GPARAM_DET_TX));
	fprintf(fh, "%i 0 0\n", mille_label(0, GPARAM_DET_TY));
	fprintf(fh, "%i 0 %i\n", mille_label(0, GPARAM_DET_TZ), refine_clen ? 0 : -1);
	fprintf(fh, "%i 0 %i\n", mille_label(0, GPARAM_DET_RX), out_of_plane_tilts ? 0 : -1);
	fprintf(fh, "%i 0 %i\n", mille_label(0, GPARAM_DET_RY), out_of_plane_tilts ? 0 : -1);
	fprintf(fh, "%i 0 -1\n", mille_label(0, GPARAM_DET_RZ));
	fprintf(fh, "%i 0 0\n", mille_label(0, GPARAM_SCANV_BEAM_A));
	fprintf(fh, "%i 0 0\n", mille_label(0, GPARAM_SCANV_BEAM_B));
	fprintf(fh, "%i 0 0\n", mille_label(0, GPARAM_SCANV_BEAM_C));
	fprintf(fh, "%i 0 0\n", mille_label(0, GPARAM_SCANV_BEAM_D));

	for ( i=0; i<n_groups; i++ ) {
		int f_inplane = (groups[i].hierarchy_level > level) ? -1 : 0;
		int f_outplane_shift = out_of_plane_shift ? f_inplane : -1;
		int f_outplane_tilts = out_of_plane_tilts ? f_inplane : -1;
		if ( groups[i].hierarchy_level == 0 ) continue;
		fprintf(fh, "%i 0 %i\n", mille_label(groups[i].serial, GPARAM_DET_TX), f_inplane);
		fprintf(fh, "%i 0 %i\n", mille_label(groups[i].serial, GPARAM_DET_TY), f_inplane);
		fprintf(fh, "%i 0 %i\n", mille_label(groups[i].serial, GPARAM_DET_TZ), f_outplane_shift);
		fprintf(fh, "%i 0 %i\n", mille_label(groups[i].serial, GPARAM_DET_RX), f_outplane_tilts);
		fprintf(fh, "%i 0 %i\n", mille_label(groups[i].serial, GPARAM_DET_RY), f_outplane_tilts);
		fprintf(fh, "%i 0 %i\n", mille_label(groups[i].serial, GPARAM_DET_RZ), f_inplane);
	}
	fprintf(fh, "\n");

	/* All corrections must sum to zero at each level of hierarchy */
	if ( make_zero_sum(fh, groups, n_groups, "all", level,
	                   out_of_plane_shift, out_of_plane_tilts) ) return 1;

	if ( big_errors ) {
		fprintf(fh, "hugecut 1000\n");
	}

	fprintf(fh, "method inversion 5 0.1\n");
	fprintf(fh, "closeandreopen\n");
	fprintf(fh, "entries %i\n", min_measurements);
	fprintf(fh, "end\n");
	fclose(fh);

	unlink("millepede.res");

	if ( run_pede() ) return 1;
	STATUS("Millepede succeeded.\n\n");

	fh = fopen("millepede.res", "r");
	if ( fh == NULL ) {
		ERROR("Failed to open millepede.res\n");
		return 1;
	}

	if ( fgets(line, 256, fh) != line ) {
		ERROR("Failed to read first line of millepede.res\n");
		return 1;
	}
	if ( strncmp(line, " Parameter ", 11) != 0 ) {
		ERROR("First line of millepede.res is not as expected.\n");
		return 1;
	}

	data_template_reset_total_movements(dtempl);

	int last_group_serial = -1;
	do {

		char **bits;
		int i, n;

		rval = fgets(line, 256, fh);
		if ( rval != line ) continue;

		chomp(line);
		notrail(line);
		n = assplode(line, " ", &bits, ASSPLODE_NONE);
		if ( (n != 3) && (n != 5) ) {
			ERROR("Didn't understand this line from Millepede: (%i) %s", n, line);
			return 1;
		}

		if ( n == 5 ) {

			int code;
			double shift;
			int p;
			const char *group_name;
			int group_serial;
			double updates[4] = {0,0,0,0};


			if ( convert_int(bits[0], &code) ) {
				ERROR("Didn't understand '%s'\n", bits[0]);
				return 1;
			}
			if ( convert_float(bits[1], &shift) ) {
				ERROR("Didn't understand '%s'\n", bits[1]);
				return 1;
			}

			p = mille_unlabel(code % 100);
			group_serial = code - (code % 100);
			group_name = group_serial_to_name(group_serial,
			                                  groups,
			                                  n_groups);

			if ( last_group_serial != group_serial ) {
				STATUS("Group %s:\n", group_name);
				last_group_serial = group_serial;
			}

			switch ( p ) {
				case GPARAM_DET_TX:
				case GPARAM_DET_TY:
				case GPARAM_DET_TZ:
				STATUS("   %14s %+f mm\n", str_param(p), 1e3*shift);
				break;

				case GPARAM_DET_RX:
				case GPARAM_DET_RY:
				case GPARAM_DET_RZ:
				STATUS("   %14s %+f deg\n", str_param(p), rad2deg(shift));
				break;

				case GPARAM_SCANV_BEAM_A:
				case GPARAM_SCANV_BEAM_B:
				case GPARAM_SCANV_BEAM_C:
				case GPARAM_SCANV_BEAM_D:
				STATUS("   %14s %+f mm/scan unit\n", str_param(p), 1e3*shift);
				break;
			}

			if ( group_name == NULL ) {
				ERROR("Invalid group serial number %i\n", code);
				return 1;
			}

			switch ( p ) {

				case GPARAM_DET_TX:
				data_template_translate_group_m(dtempl, group_name,
				                                -shift, 0, 0);
				break;

				case GPARAM_DET_TY:
				data_template_translate_group_m(dtempl, group_name,
				                                0, -shift, 0);
				break;

				case GPARAM_DET_TZ:
				data_template_translate_group_m(dtempl, group_name,
				                                0, 0, -shift);
				break;

				case GPARAM_DET_RX:
				data_template_rotate_group(dtempl, group_name,
				                           -shift, 'x');
				break;

				case GPARAM_DET_RY:
				data_template_rotate_group(dtempl, group_name,
				                           -shift, 'y');
				break;

				case GPARAM_DET_RZ:
				data_template_rotate_group(dtempl, group_name,
				                           -shift, 'z');
				break;

				case GPARAM_SCANV_BEAM_A:
				updates[0] = -shift;
				data_template_add_to_scanv_beamxy(dtempl, updates);
				break;

				case GPARAM_SCANV_BEAM_B:
				updates[1] = -shift;
				data_template_add_to_scanv_beamxy(dtempl, updates);
				break;

				case GPARAM_SCANV_BEAM_C:
				updates[2] = -shift;
				data_template_add_to_scanv_beamxy(dtempl, updates);
				break;

				case GPARAM_SCANV_BEAM_D:
				updates[3] = -shift;
				data_template_add_to_scanv_beamxy(dtempl, updates);
				break;

				default:
				ERROR("Invalid parameter %i\n", p);
				return 1;
			}

		}

		for ( i=0; i<n; i++ ) free(bits[i]);
		free(bits);

	} while ( rval == line );

	fclose(fh);

	if ( panel_totals ) {
		data_template_print_total_movements(dtempl);
	}

	data_template_write_to_file(dtempl, out_geom);

	if ( warn_times ) print_time_warning();

	return 0;
}
